test/e2e/storage/testsuites/provisioning.go

```diff
@@ -269,6 +269,15 @@ func (p *provisioningTestSuite) DefineTests(driver storageframework.TestDriver,
 			}
 			e2evolume.TestVolumeClientSlow(f, testConfig, nil, "", tests)
 		}
+		// For GCE PD, cloning fails if the source disk has another operation going on. When we try create the clone in TestDynamicProvisioning,
+		// the source disk could still be in the process of detaching since InjectContent in preparePVCDataSourceForProvisioning defers
+		// deleting the injector pod. To verify the disk is ready, we create a pod with the source claim and wait for the pod to be ready.
+		source, _ := l.cs.CoreV1().PersistentVolumeClaims(l.sourcePVC.Namespace).Create(context.TODO(), l.sourcePVC, metav1.CreateOptions{})
+		pod, _ := e2epod.CreatePod(l.cs, source.Namespace, nil /* nodeSelector */, []*v1.PersistentVolumeClaim{source}, true /* isPrivileged */, "while true ; do sleep 2; done" /* command */)
+		defer func() {
+			e2epod.DeletePodOrFail(l.testCase.Client, pod.Namespace, pod.Name)
+			e2epod.WaitForPodToDisappear(l.testCase.Client, pod.Namespace, pod.Name, labels.Everything(), framework.Poll, f.Timeouts.PodDelete)
+		}()
```

Comment:
Is this chunk still necessary? It looks like a GCE PD specific workaround, which we're trying to avoid in generic CSI tests.

Reply:
The test has a race condition where the cloning begins while the source disk is still being created. This causes extra reconcile loops that causes the test to timeout. I suppose other kinds of disks that clone aren't susceptible to this race?

The CSI spec doesn't seem to say one way or another what the behavior should be for multiple pending operations on a volume (ie, in this case, controller unpublish and cloning). But it seems reasonable to me that tests should avoid such kinds of race conditions. We also know that trying to work around them by adding timeouts is unreliable and leads to flakey tests.

We can change the comment so that it's not GCE-specific. It might also work to change the source creation so that volume deletion is not deferred, but is explicitly blocked on. That requires deeper surgery down in the helper functions, though.

Reply:
Before this PR, it tested volume cloning with the source volume unmounted (possibly conflicting with its detach). Now it tests cloning of a mounted volume. It's a good test to have, still, it's something else than the original one.

Looking at it from user perspective. When I clone a volume at a wrong time, it can conflict with volume detach and extra reconciliation may be needed in the driver / external-provisioner. How many reconciliations are expected here? The provisioning timeout in the test is 5 minutes and you can increase it to anything that works for GCE. But it is the time that some of your users will experience when cloning volumes. They won't run a pod before or wait for a volume to get fully detached.

Reply:
A test for cloning while the volume is still attached is a good idea. A test for cloning while a detach is occurring could be a good idea, although difficult to get to run consistently. But a test that races and sometimes tests detached and sometimes attached because we're lazy about a deferred pod deletion is just flaky.

I feel quite strongly that we should eliminate race conditions in tests to make them behave as predictably as possible. Tests on the scale of the e2e ones are difficult enough to deflake with all their infrastructure dependencies, I think we should make sure the tests themselves are as deterministic as possible.

Reply:
I agree that the current approach is flaky. Would it be possible to wait for a volume to get fully detached before cloning the volume? It's would be reliable to wait for VolumeAttachment to disappear, but in-tree volume plugins don't have it and it's hard to get the attachment ID.

I fear that cloning of a mounted volume will be flaky because mount of the clone may not succeed, because the filesystem is corrupted. It should not happen with ext4 / xfs, still, it's a possibility with other filesystems.

Reply:
That's a good point of the problems of cloning a mounted volume. This was our solution to making sure the volume was in a consistent state, but we would have in fact preferred just knowing when it was detached.

Maybe we could look at the volumesInUse on the nodes? The test clusters aren't big, so it should be scalable to just get all the nodes and wait until the volume doesn't appear on any of them (?)

Reply:
We have the similar issue of online vs offline with snapshots too. We had recently turned our online snapshot tests to offline due to Windows data consistency issues. But I think that we actually test both. For the online consistency case I think the tests should do a data flush, which wasn't available in our Windows test images before but are available now.
https://github.com/kubernetes/kubernetes/issues/101333

Regarding waiting for detach, we only support cloning with CSI volumes so it may ok to look at VolumeAttachments.

Reply:
> Regarding waiting for detach, we only support cloning with CSI volumes so it may ok to look at VolumeAttachments.

Good point! @amacaskill, can you please add waiting for VolumeAttachment to be removed before the test starts cloning? There already is code that calculates VolumeAttachment name, perhaps you could create a function for it: https://github.com/kubernetes/kubernetes/blob/bca1b79728888b307f7ddaf40a60f53c71adbaec/test/e2e/storage/csi_mock_volume.go#L377-L379

And code that waits for VolumeAttachment to be removed:
https://github.com/kubernetes/kubernetes/blob/bca1b79728888b307f7ddaf40a60f53c71adbaec/test/e2e/storage/csi_mock_volume.go#L2087

Reply:
Yes I'll add that in

Reply:
I added it in. The only problem was that for the provisioning tests, ClientNodeSelection.Name is "" and the default node is any node, so we don't know the node name to be able to generate this hash  of the volumeHandle, provisioner, and ClientNodeSelection.Name. See the workaround I did instead [here](https://github.com/kubernetes/kubernetes/blob/8102bbe05a0d487db04b088199ff7cd029f9aca5/test/e2e/framework/volume/fixtures.go#L244-L261). Is this ok?